fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/RecordQueryInUnionOnValuePlan.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * RecordQueryUnionOnKeyExpressionPlan.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2022 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.query.plan.plans;
+
+import com.apple.foundationdb.record.RecordCoreException;
+import com.apple.foundationdb.record.metadata.expressions.KeyExpression;
+import com.apple.foundationdb.record.query.plan.cascades.AliasMap;
+import com.apple.foundationdb.record.query.plan.cascades.CorrelationIdentifier;
+import com.apple.foundationdb.record.query.plan.cascades.GroupExpressionRef;
+import com.apple.foundationdb.record.query.plan.cascades.Quantifier;
+import com.apple.foundationdb.record.query.plan.cascades.values.Value;
+import com.google.common.collect.Iterables;
+
+import javax.annotation.Nonnull;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Function;
+
+/**
+ * Union plan that compares using a {@link Value}.
+ */
+public class RecordQueryInUnionOnValuePlan extends RecordQueryInUnionPlan {
+    @Nonnull
+    private final CorrelationIdentifier baseAlias;
+
+    public RecordQueryInUnionOnValuePlan(@Nonnull final Quantifier.Physical inner,
+                                         @Nonnull final List<? extends InSource> inSources,
+                                         @Nonnull final Function<CorrelationIdentifier, Value> comparisonKeyValueFunction,
+                                         final boolean reverse,
+                                         final int maxNumberOfValuesAllowed) {
+        this(inner,
+                inSources,
+                CorrelationIdentifier.uniqueID(),
+                comparisonKeyValueFunction,
```

Comment:
This is a strange contract, that can only work if `comparisonKeyValueFunction` ignores its argument. Why doesn't this just take the `Value` directly?